src/ipa_pc/mod.rs

```diff
@@ -736,6 +736,136 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
         })
     }
 
+    /// Refers to the "Multi-poly multi-point assertions" section of the "Recursive SNARKs based on Marlin" document
+    fn batch_open_individual_opening_challenges<'a>(
+        ck: &Self::CommitterKey,
+        labeled_polynomials: impl Clone + IntoIterator<Item = &'a LabeledPolynomial<G::ScalarField>>,
+        commitments: impl Clone + IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
+        query_set: &QuerySet<G::ScalarField>,
+        opening_challenges: &dyn Fn(u64) -> G::ScalarField,
+        rands: impl Clone + IntoIterator<Item = &'a Self::Randomness>,
+        rng: Option<&mut dyn RngCore>,
+    ) -> Result<Self::BatchProof, Self::Error>
+    where
+        Self::Randomness: 'a,
+        Self::Commitment: 'a,
+    {
+        let batch_time = start_timer!(|| "Multi poly multi point batching.");
+
+        let polys_iter = labeled_polynomials.clone().into_iter();
+
+        let mut opening_challenge_counter = 0;
+        
+        // lambda
+        let mut cur_challenge = opening_challenges(opening_challenge_counter);
+        opening_challenge_counter += 1;
+
+        let mut query_to_labels_map = BTreeMap::new();
+
+        for (label, (point_label, point)) in query_set.iter() {
+            let labels = query_to_labels_map
+                .entry(point_label)
+                .or_insert((point, BTreeSet::new()));
+            labels.1.insert(label);
+        }
+
+        let mut points = vec![];
+
+        // h(X)
+        let mut batch_polynomial = Polynomial::zero();
+
+        for (
+            (_point_label, (point, _labels)), 
+            labeled_polynomial
+        ) in query_to_labels_map.into_iter().zip(polys_iter) {
```

Comment:
I'll unresolve this just to keep track of it (maybe we can open a separate issue). The prover is quite slow, and it's highly affected also by this portion of computation.
We need to parallelize the underlying polynomial arithmetic ( https://github.com/HorizenOfficial/ginger-lib/issues/87 ), use the Horner scheme to evaluate the polynomial (arkworks has a function horner_evaluate) and other optimizations to the underlying arithmetic, see the performances and compare them by doing everything in serial and parallelize the for loop instead.
This is not urgent for now but it will be in the future.